### DISCUSSION TITLE
Is using the inverter's LAN port an option in the future?

### DISCUSSION THREAD

**Discussion by andrewmooreio** [nathanmarlor/foxess_em#224] [General]
Hi @nathanmarlor.

I currently have Energy Management setup using the FoxESS Cloud integration, however it's proven to be fairly unreliable. For example, my battery didn't charge last night because the FoxESS API was unavailable, meaning I'm running off the grid this morning until the sun comes up. I also use [Chris' FoxESS Modbus integration](https://github.com/StealthChesnut/HA-FoxESS-Modbus) via direct LAN into the inverter, and it's far more reliable.

- I see you recently added Modbus support to EM, but the readme mentions the direct LAN isn't supported. Is this something that will be possible in the future? If not, I'll plan to install a W610.
- Slight tangent, but I see you have also written your own [FoxESS Modbus integration](https://github.com/nathanmarlor/foxess_modbus) recently, how does this compare to the one linked above?

Many thanks in advance.

**Comment by nathanmarlor** [comment]
Hi @andrewmooreio 

Interesting that it didn't charge - I recently inverted the logic so that the charge periods are set on permanently, and turned off if not required at the start of the off-peak. This was done to 'fail safe' and ensure a full charge upon failure of the Fox cloud. Happy to take a look at last night's logs if you have them available (settings -> system -> logs). Can you also include a screenshot of the devices overview page showing average load etc.

In terms of modbus integration vs. the original it was written to improve performance and upgrades. The original integration reads ~40 registers sequentially in a ~5 second poll which eventually overloads HA and causes some nasty hangs on startup/shutdown. The new integration reads 8 registers at a time (that's all the inverter will allow), which significantly improves things. Upgrades are now controlled through HACS semi-automatically which means new registers, improved logic etc are delivered without needing to wrangle yaml etc. 

Hope this helps!

**Reply by Captain0Kangaroo** [reply]
Sorry to jump in here but I'm currently using the original modbus version and would like to give yours a go and see how it works out.

2 questions:

Is it ok to run them side by side to test it out?

And if I make the switch is it possible to retain all my data by using the same sensors etc? I really don't want to loose my energy data.

**Reply by nathanmarlor** [reply]
@Captain0Kangaroo I wouldn't advise running them side by side as you'll be overloading the W610 even further. The easiest migration route is to comment out the modbus configuration entries, reboot HA, confirm your sensors show as below (the native modbus integration has released the sensor - expect a different name) then follow the standard HACS install. The sensors are named identically so once the new integration fires up all of your history etc. will be retained. 

![image](https://user-images.githubusercontent.com/3436911/226878408-aed40236-01cd-4bbc-ac31-8c7dcffc6ad2.png)

**Reply by andrewmooreio** [reply]
Thanks a lot @nathanmarlor. I may have been too quick to blame FoxESS for the failure last night, as the API error might not quite line up with the charging window (00:30-04:30). Looking at the calendar, it shows 0 kWh scheduled for charging last night. So it looks like the prediction was off instead of a failure to charge. I'm on 1.6.3, yet to update to 1.7.0.

![image](https://user-images.githubusercontent.com/20435317/226887967-a4beee9e-622d-4252-bcbf-40d2ce5faddf.png)
![image](https://user-images.githubusercontent.com/20435317/226888079-f4c05b42-59fa-4b3d-ae63-39f99914587b.png)
![image](https://user-images.githubusercontent.com/20435317/226888171-f1f1d42e-59bf-44ec-a06a-449c21142107.png)
![image](https://user-images.githubusercontent.com/20435317/226888237-c64690ff-a1fa-4702-94eb-0f952bd04048.png)

**Reply by nathanmarlor** [reply]
Ah, I think I know what's going on here. You're not on the latest inverter version? It fails to set the charge current through the cloud so then fails to set min SoC etc subsequently. I fixed this in v1.7.0 :)

**Reply by andrewmooreio** [reply]
My inverter firmware versions are as follows:

- Master : 1.54
- Slave : 1.02
- Manager : 1.57

Not sure what the latest is, but it was updated not that long ago. I'll make sure to upgrade EM to v1.7.0.

**Comment by nathanmarlor** [comment]
Wrt LAN registers - it's unlikely these will be discovered any time soon so I'd advise going down the 485 route. Interestingly the performance of the W610 has been proven to be a little sketchy of late, others have had some great success with the Waveshare devices -> https://thepihut.com/products/industrial-serial-server-rs485-to-rj45-ethernet

I've just ordered the PoE version from their site and am awaiting delivery. @FozzieUK knows more :)

**Reply by FozzieUK** [reply]
@Captain0Kangaroo If you upgrade to the latest version of the stealthchestnut RS485 modbus, the performance drops off a cliff, randomly blocking HA on startup and shutdown. @nathanmarlor version uses a lot less resources and the blocking stopped as soon as I installed it.
I’ve done testing using the W610 which seems to be most peoples go to TCP>485 converter and it really struggles to keep up - a complete poll taking nearly 6 seconds. Nathan & Anthony have done a great job with the modbus integration, improving the performance and it’s now nearly twice as fast as it was.
The converter in the link above is very fast, I would strongly recommend if you’ve got the LAN cable in situ already, and using that instead of the W610.
Main recommendation for installing the new integration is to comment out the old include modbuslan.yaml first, then restart so that all the old sensors are orphaned from it - then install the new integration and it will take over the old sensors so your statistics have continuity.

**Reply by andrewmooreio** [reply]
Thanks, @nathanmarlor. I've ordered one of the Waveshare converters.

**Reply by Captain0Kangaroo** [reply]
@FozzieUK

Just tried what you suggested (comment out the old include modbuslan.yaml) and then installed via HACS but non of my data worked: 

Any ideas?
 
<img width="1455" alt="Screenshot 2023-03-22 at 17 34 24" src="https://user-images.githubusercontent.com/96144685/226991555-86f6d539-5afc-46a7-b4c1-8cdea6eec8e4.png">

**Reply by nathanmarlor** [reply]
@Captain0Kangaroo have they appeared as _2 by chance? I think you may need to fly through and rename them back.

**Reply by Captain0Kangaroo** [reply]
No, it looks like their different altogether for example:

sensor.fox_modbus_pv1_power

Vs

sensor.pv1_power

**Reply by FozzieUK** [reply]
I’ve never seen that before - as long as the old modbus is no longer providing any sensors, it’s a bit of a pain to do it but if you go to the fox modbus integration and select the ‘1 device’ then click on each sensor and modify it’s name (removing the ‘fox_modbus_’) and click update, then repeat on the next.
I had to do it on a system this morning, only takes a few mins to do them all.

**Reply by Captain0Kangaroo** [reply]
I'm not actually using the Fox Modbus HACS install. It was a manual install when it first came out in May last year so not sure if that's the issue?

**Reply by nathanmarlor** [reply]
@Captain0Kangaroo sounds like that's the issue, as even the other repo doesn't have the fox_modbus prefix anymore.

**Reply by Captain0Kangaroo** [reply]
I'm guessing my only option is to stay as I am or rip out the old one and start again loosing my data?

**Reply by nathanmarlor** [reply]
@Captain0Kangaroo you could try setting the 'friendly name' to 'fox_modbus', this will prefix all sensors with that.

**Reply by nathanmarlor** [reply]
> @FozzieUK
> 
> Just tried what you suggested (comment out the old include modbuslan.yaml) and then installed via HACS but non of my data worked: 
> 
> Any ideas?
>  
> <img width="1455" alt="Screenshot 2023-03-22 at 17 34 24" src="https://user-images.githubusercontent.com/96144685/226991555-86f6d539-5afc-46a7-b4c1-8cdea6eec8e4.png">
> 
> 

Just looked closer at this - the sensors aren't prefixed with fox_modbus?

**Reply by Captain0Kangaroo** [reply]
> > @FozzieUK
> > 
> > Just tried what you suggested (comment out the old include modbuslan.yaml) and then installed via HACS but non of my data worked: 
> > 
> > Any ideas?
> >  
> > <img width="1455" alt="Screenshot 2023-03-22 at 17 34 24" src="https://user-images.githubusercontent.com/96144685/226991555-86f6d539-5afc-46a7-b4c1-8cdea6eec8e4.png">
> > 
> > 
> 
> Just looked closer at this - the sensors aren't prefixed with fox_modbus?

No... this was a beta version that I installed back in May before it was launched so sensors are named differently.

I'm a bit stuck 🙈

**Reply by andrewmooreio** [reply]
Hi @FozzieUK. Would you mind sharing your Waveshare RS485 > RJ45 settings please? Mine has arrived and I can't get it to work.

Testing with modpoll I get a timeout:

```
./modpoll -m tcp -p 4196 -a 247 -t 3 -0 -1 -c 16 -r 11000 172.16.1.36
modpoll 3.10 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2021 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC4
Slave configuration...: address = 247, start reference = 11000 (PDU), count = 16
Communication.........: 172.16.1.36, port 4196, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, input register table

-- Polling slave...
Reply time-out!
```
The Waveshare is running firmware v1.452. I've upped the instruction timeout to 5 seconds, hoping that would help, but it didn't.
![image](https://user-images.githubusercontent.com/20435317/230144176-b85692ac-f502-4dde-92bd-c0309a4d5a9b.png)

Thanks in advance.

**Comment by FozzieUK** [comment]
@andrewmooreio sure, here you go - I use default modbus port of 502

![image](https://user-images.githubusercontent.com/113460294/230147275-776e1ea6-66a1-4cbe-a27b-bee41484f1e6.png)

**Reply by FozzieUK** [reply]
I also used the Vir Com application which also lets you get to the advanced settings (although I think the webpage might default these)

![image](https://user-images.githubusercontent.com/113460294/230148171-f0e89e4d-1405-4c69-a236-f123cbdc751a.png)
![image](https://user-images.githubusercontent.com/113460294/230148255-131f3d37-518b-46b2-9100-f51a76dbd160.png)

**Reply by andrewmooreio** [reply]
Thanks! Did you change anything in the advanced settings via Vircom? I've matched your settings exactly through the web interface, and I'm still getting a timeout.

Perhaps it is dodgy RS485 wiring, I'll check and report back.

**Comment by FozzieUK** [comment]
@andrewmooreio - the Inverter has to be connected via RS485 or it will reply timeout - is this the first time you've connected your inverter by RS485? - if so you must make sure the RS485A goes to RS485A and 485B to 485B, a common mistake is to wire it to the wrong pins should be pins 3 & 4 (not 1 & 2)
![image](https://user-images.githubusercontent.com/113460294/230149596-96951882-d0b5-4e6a-9409-69dc6c1de427.png)

**Reply by andrewmooreio** [reply]
'Twas my dodgy wiring, thanks for the suggestion. Everything was in the right pins, just not seated properly.
